File: examples/typescript/workload-scaling-policy-example.ts
import * as pulumi from "@pulumi/pulumi";
import * as castai from "@castai/pulumi-castai";

// Initialize the CAST AI provider
const provider = new castai.Provider("castai", {
    // API token will be read from the environment variable CASTAI_API_TOKEN
});

// Reference an existing EKS cluster (you can use any cluster type)
const eksCluster = new castai.aws.EksCluster("example-eks-cluster", {
    accountId: "123456789012", // Replace with your AWS account ID
    region: "us-west-2",       // Replace with your AWS region
    eksClusterName: "my-eks-cluster", // Replace with your EKS cluster name
    securityGroupId: "sg-12345678", 
    subnetIds: ["subnet-12345678", "subnet-87654321"],
});

// Create a workload scaling policy
const scalingPolicy = new castai.workload.ScalingPolicy("example-scaling-policy", {
    clusterId: eksCluster.id,
    name: "web-services-scaling",
    enabled: true,
    
    // Define the scope of the policy
    namespaceSelector: {
        matchLabels: {
            "environment": "production"
        }
    },
    
    // Configure the metrics for scaling
    metrics: [
        {
            type: "cpu",
            weight: 70,
            stableWindow: "5m",
            threshold: {
                utilizationPercentage: 80
            }
        },
        {
            type: "memory",
            weight: 30,
            stableWindow: "5m",
            threshold: {
                utilizationPercentage: 75
            }
        }
    ],
    
    // Configure vertical scaling
    verticalScaling: {
        enabled: true,
        vpa: {
            updateMode: "Auto",
            minAllowed: {
                cpu: "100m",
                memory: "128Mi"
            },
            maxAllowed: {
                cpu: "4",
                memory: "8Gi"
            },
            controlledResources: ["cpu", "memory"]
        }
    },
    
    // Configure horizontal scaling
    horizontalScaling: {
        enabled: true,
        hpa: {
            minReplicas: 2,
            maxReplicas: 10,
            behavior: {
                scaleDown: {
                    stabilizationWindowSeconds: 300,
                    policies: [
                        {
                            type: "Percent",
                            value: 10,
                            periodSeconds: 60
                        }
                    ]
                },
                scaleUp: {
                    stabilizationWindowSeconds: 0,
                    policies: [
                        {
                            type: "Percent",
                            value: 100,
                            periodSeconds: 15
                        }
                    ]
                }
            }
        }
    }
});

// Export the policy ID
export const policyId = scalingPolicy.id; 

File: examples/python/workload_scaling_policy_example.py
"""
CAST AI Workload Scaling Policy Example for Python
This example demonstrates how to configure workload scaling policies for a CAST AI cluster.
"""

import pulumi
import pulumi_castai as castai
import os
from models import (
    Provider as ProviderModel,
    ScalingPolicy,
    NamespaceSelector,
    ScalingPolicyMetric, 
    MetricThreshold,
    VerticalScaling,
    VerticalScalingVpa,
    VerticalScalingVpaMinAllowed,
    VerticalScalingVpaMaxAllowed,
    HorizontalScaling,
    HorizontalScalingHpa,
    HorizontalScalingHpaBehavior,
    HorizontalScalingHpaBehaviorScaleDown,
    HorizontalScalingHpaBehaviorScaleUp,
    HorizontalScalingHpaBehaviorPolicy
)

# Initialize the CAST AI provider
api_token = os.environ.get("CASTAI_API_TOKEN", "your-api-token-here")
provider_config = ProviderModel(api_token=api_token)
provider = castai.Provider("castai-provider",
    api_token=provider_config.api_token,
    api_url=provider_config.api_url
)

# Placeholder: Assume cluster_id is obtained from an existing cluster resource
# e.g., eks_cluster = castai.aws.EksCluster(...)
# cluster_id = eks_cluster.id
cluster_id = "your-cluster-id" # Replace with a valid cluster ID

# Create a workload scaling policy using Pydantic models
scaling_policy_config = ScalingPolicy(
    cluster_id=cluster_id,
    name="web-services-py-scaling",
    enabled=True,
    namespace_selector=NamespaceSelector(
        match_labels={
            "environment": "production"
        }
    ),
    metrics=[
        ScalingPolicyMetric(
            type="cpu",
            weight=70,
            stable_window="5m",
            threshold=MetricThreshold(
                utilization_percentage=80
            )
        ),
        ScalingPolicyMetric(
            type="memory",
            weight=30,
            stable_window="5m",
            threshold=MetricThreshold(
                utilization_percentage=75
            )
        ),
    ],
    vertical_scaling=VerticalScaling(
        enabled=True,
        vpa=VerticalScalingVpa(
            update_mode="Auto",
            min_allowed=VerticalScalingVpaMinAllowed(
                cpu="100m",
                memory="128Mi"
            ),
            max_allowed=VerticalScalingVpaMaxAllowed(
                cpu="4",
                memory="8Gi"
            ),
            controlled_resources=["cpu", "memory"]
        )
    ),
    horizontal_scaling=HorizontalScaling(
        enabled=True,
        hpa=HorizontalScalingHpa(
            min_replicas=2,
            max_replicas=10,
            behavior=HorizontalScalingHpaBehavior(
                scale_down=HorizontalScalingHpaBehaviorScaleDown(
                    stabilization_window_seconds=300,
                    policies=[
                        HorizontalScalingHpaBehaviorPolicy(
                            type="Percent",
                            value=10,
                            period_seconds=60
                        )
                    ]
                ),
                scale_up=HorizontalScalingHpaBehaviorScaleUp(
                    stabilization_window_seconds=0,
                    policies=[
                        HorizontalScalingHpaBehaviorPolicy(
                            type="Percent",
                            value=100,
                            period_seconds=15
                        )
                    ]
                )
            )
        )
    )
)

# Convert Pydantic model to arguments dictionary
scaling_policy_args = scaling_policy_config.dict(exclude_none=True)

# Create the scaling policy resource using the direct type
scaling_policy = castai.ScalingPolicy(
    "example-py-scaling-policy",
    **scaling_policy_args,
    opts=pulumi.ResourceOptions(provider=provider)
)

# Export the policy ID
pulumi.export("policy_py_id", scaling_policy.id) 